### PR TITLE
Sample types

### PR DESCRIPTION
**Pull request by marikoja**
Sample types from hamburger menu
View, create, edit, delete.

### CODE REVIEW

File: frontend/src/helpers/API.js [hunk 1]
```diff
@@ -1,85 +1,65 @@
...
 
-      if (status === 401 && data.error !== 'Invalid') {
-        setLoginOutError(data.error);
+// TODO: REPLACE ALL ALERTS WITH PROPER HANDLING
+/* Use interceptors to handle error responses generalized by status code
+   We can customize the response using the pathname */
+axiosInstance.interceptors.response.use(
+  (response) => response,
+  // eslint-disable-next-line consistent-return
+  (error) => {
```
Comment (gnomicosuw) [inline frontend/src/helpers/API.js:34]: this catch does not appear to work.  I suggest something like as below
NOTES: 
- I just put in an alert (we need to write the code to handle the actual error
- If the system hits an error here it still needs to pass something back. I suggest simply passing back "false"
- I also included "suggested usage" section (based on what I started for object_types)

SUGGESTED CODE:
  (error) => {
    const error_response = error.response;
    switch (error_response["status"]) {
      case 400:
        alert("400: "+JSON.stringify(error_response["data"]));
        break;
      case 401:
        if (window.location.pathname !== '/login') {
          if (currentSessionToken) {
            axios.post('/token/delete');
            sessionStorage.clear('token');
          }
        }
        /* TODO: HANDLE SESSION TIMEOUT
           if (...pathname !== '/login' && message !== 'Session timeout') { OPEN LOGIN MODAL} */
        alert("401: "+JSON.stringify(error_response["data"]));
        break;
      case 403:
        // TODO: HANDLE PERMISSIONS
        alert("403: "+JSON.stringify(error_response["data"]));
        break;
      case 404:
        alert("404: "+JSON.stringify(error_response["data"]));
        break;
      default:
        alert(error_response["status"]+": "+JSON.stringify(error_response["data"]));
    }
    // Return something obvious to stop processing. I like simply "false"
    return false;
  },

SUGGESTED USAGE 
   ...
    // make an API call    
    await objectsAPI.create(formData);

    // break if the HTTP call resulted in an error (the "return false" above)
    // NOTE: the alert("break") is just there for testing. Whatever processing should be handled in API.js, and we just need stop the system from trying to continue...
    if (!response) {
      alert("break") 
      return;
    }

    // process errors (this is defined by our API architecture, we <either> return a success <or> { "errors" => ... }
    // NOTE: in this example I am processing by putting up an AlertToast
    const errors = response["errors"];
    if (errors) {
      return setAlertProps({
        message: JSON.stringify(errors, null, 2),
        severity: 'error',
        open: true,
      });
    }

    // success
    ... continue processing

File: frontend/src/components/app/App.jsx [hunk 2]
```diff
@@ -59,7 +77,11 @@ export default function App() {
 
             {/* Left Hamburger Menu */}
             <Route exact path="/users" render={(props) => <UsersPage {...props} />} />
+
             <Route exact path="/sample_types" render={(props) => <SampleTypesPage {...props} />} />
+            <Route exact path="/sample_types/new" render={(props) => <SampleTypeDefinitionForm {...props} />} />
+            <Route exact path="/sample_types/:id/edit" render={(props) => <SampleTypeDefinitionForm {...props} />} />
+
```
Comment (gnomicosuw) [inline frontend/src/components/app/App.jsx:84]: After all the <Route... /> elements we should redirect any other paths somewhere.  Right now the system just puts up a blank page with a header, which is really weird. (This is the closest place I could add this comment.)

File: frontend/src/components/sampleTypes/SampeTypeDefinitionForm.jsx [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+  useEffect(() => {
+    const fetchDataById = async () => {
+      const data = await samplesAPI.getTypeById(match.params.id);
+      setFieldTypes(data.field_types);
+      setSampleTypeDescription(data.description);
+      setSampleTypeName(data.name);
+      setObjectTypes(data.object_types);
+      setInventory(data.inventory);
+      setId(data.id);
+      setIsLoading(false);
```
Comment (gnomicosuw) [inline frontend/src/components/sampleTypes/SampeTypeDefinitionForm.jsx:96]: this line is not needed.  Every page ("new" vs "edit") calls await samplesAPI.getTypes(); (line 79) so the only call that we need to turn off the spinner is setIsLoading(false); on line 81. 

p.s., we had also talked about using 2 different calls, and making sure only one gets called on each page type. I don't have a strong preference either way.

File: frontend/src/components/sampleTypes/SampeTypeDefinitionForm.jsx [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    };
+
+    match.params.id ? fetchDataById() : '';
+  }, []);
+
+  /*  Update allowSubmit state if name and Description change
+      Disable submit if name or description are empty */
+  useEffect(() => {
+    setDisableSubmit(!(!!sampleTypeName || !!sampleTypeDescription));
```
Comment (gnomicosuw) [inline frontend/src/components/sampleTypes/SampeTypeDefinitionForm.jsx:105]: this logic should be 
`setDisableSubmit(!sampleTypeName.trim() || !sampleTypeDescription.trim());`

Note that on the backend both name and description are required (and whitespace is truncated), so if either of the two fields are "empty" then the system will return a form error.  Catching it on the front-end is a nice addition to the UX, and the logic above is more accurate.

**Comment by gnomicosuw** [comment]
<AlertToast /> does not work as expected.

Once an AlertToast is shown and closed, the AlertToast shows up again every time the user types in an <input> area, which is very confusing and very distracting.

Repro:
1. navigate to a page with a form  
NOTE: (you can use the new sample type form but that page has other errors which makes it confusing. I can demo on a different page in a different branch)

2. submit the form, but with errors
result: you get the <AlertToast />

3. close the toast

4. make any change to any input field (such as a text field)
result: the <AlertToast /> shows up as soon as you type in the field.

NOTE:  the useEffect code below gets triggered and reopens the toast as soon as you type in an input field (you can repro this with a simple debugger statement).  I'm not sure why. 

```
  // Set state from props
  useEffect(() => {
    setMessage(props.message);
    setSeverity(props.severity);
    setOpen(props.open);
  }, [props]);

```

**Comment by gnomicosuw** [comment]
Ben, some general comments Mariko & I have discussed that we want your input on.

1. file naming conventions.  Should files names reflect component names? For example,
- the file "ShowSampleType.jsx" defines the component  < ShowSampleType >
but
- the file "SampleTypesPage.jsx" defines the component < SampleTypeDefinitions >
- the file "samples.js" defines the component < samplesAPI >

2. attribute naming conventions. Should we try to use the same names when passing and setting attributes? For example,
- the attribute "testName" gets passed to a component where "cy-data = {testName}"   Should we use the same attribute name in both places? Maybe cyData (I think "-" will cause problems, and I think eslint will complain if we use "_")

3. passing form inputs.  In my object_types branch I have used a convention where I define "baseline" form inputs using

    // set formData
    const form = document.querySelector('form');
    const data = new FormData(form);
    const formData = Object.fromEntries(data);

and then add-on "custom" inputs that after that.  In one example,

    // add new_samples_private to formData
    formData.new_samples_private = samplesPrivate;

I think we should review that strategy as a team before we start implementing a bunch of new forms...

4. the API interceptor still just tosses javascript alerts to the user on API errors. The "actual" behavior for these errors is still TBD (do we redirect the user to the homepage? do we redirect the user to their previous page? do we show a popup and let the user re-login?, etc...). In some cases the answers to those questions have not been fully spec'd, so Mariko & I suggest just leaving the alerts there for now and dealing with all of them later.  I think the three of us agree on that, but it might be worth a quick demo at some point.

5. Mariko & I came up with a strategy to simulate the equivalent of the Rails "flash[:attribute_name]" feature when we want to show a Toast on the "next" page (such as when saving something on one page and showing "saved" on the next page).  We working code is in the "object_types" branch.  I think we should review that code and integrate it here.

6. Mariko & I think we should put the "export" command at the end of each file.  (For example, in some cases the PropTypes validators appear after the "export" command - I could go either way on that if you have a strong preference.)

**Comment by bjkeller** [comment]
On style questions, we are using the Airbnb style guide
https://github.com/airbnb/javascript/tree/master/react#airbnb-reactjsx-style-guide

On Fri, Feb 19, 2021 at 10:48 AM gnomicosuw <notifications@github.com>
wrote:

> Ben, some general comments Mariko & I have discussed that we want your
> input on.
>
>    1. file naming conventions. Should files names reflect component
>    names? For example,
>
> Pretty sure this is addressed here:
https://github.com/airbnb/javascript/tree/master/react#naming
(the answer is yes)

>
>
>
>    - the file "ShowSampleType.jsx" defines the component < ShowSampleType
>    >
>    but
>    - the file "SampleTypesPage.jsx" defines the component <
>    SampleTypeDefinitions >
>    - the file "samples.js" defines the component < samplesAPI >
>
>
>    1. attribute naming conventions. Should we try to use the same names
>    when passing and setting attributes? For example,
>
>
>    - the attribute "testName" gets passed to a component where "cy-data =
>    {testName}" Should we use the same attribute name in both places? Maybe
>    cyData (I think "-" will cause problems, and I think eslint will complain
>    if we use "_")
>
> Can you point me to some code that illustrates what you are asking about?
(and as far as names, I would point you to the style guide)

>
>
>
>    1.
>
>    passing form inputs. In my object_types branch I have used a
>    convention where I define "baseline" form inputs using
>
>    // set formData
>    const form = document.querySelector('form');
>    const data = new FormData(form);
>    const formData = Object.fromEntries(data);
>
> and then add-on "custom" inputs that after that. In one example,
>
> // add new_samples_private to formData
> formData.new_samples_private = samplesPrivate;
>
> I think we should review that strategy as a team before we start
> implementing a bunch of new forms...
>

I would have to see concrete code to have an opinion.

>
>    1.
>
>    the API interceptor still just tosses javascript alerts to the user on
>    API errors. The "actual" behavior for these errors is still TBD (do we
>    redirect the user to the homepage? do we redirect the user to their
>    previous page? do we show a popup and let the user re-login?, etc...). In
>    some cases the answers to those questions have not been fully spec'd, so
>    Mariko & I suggest just leaving the alerts there for now and dealing with
>    all of them later. I think the three of us agree on that, but it might be
>    worth a quick demo at some point.
>
> The alerts are fine until we have error handling worked out. They should
be marked with TODO comments, and add a //eslint-disable-line pragma to
suppress the no-alert warning. (please don't suppress warnings as a regular
thing)

>
>    1.
>
>    Mariko & I came up with a strategy to simulate the equivalent of the
>    Rails "flash[:attribute_name]" feature when we want to show a Toast on the
>    "next" page (such as when saving something on one page and showing "saved"
>    on the next page). We working code is in the "object_types" branch. I think
>    we should review that code and integrate it here.
>
> OK

>
>    1.
>
>    Mariko & I think we should put the "export" command at the end of each
>    file. (For example, in some cases the PropTypes validators appear after the
>    "export" command - I could go either way on that if you have a strong
>    preference.)
>
> I don't see that the style guide says anything about it, but appears to
export at the beginning of the definition.

Is this something eslint checks?


>
-- 
Ben Keller
Aquarium Project Lead
Klavins Lab
Department of Electrical and Computer Engineering
University of Washington
bjkeller@uw.edu          206-685-6726

File: legacy/components/anemone/test/dummy/config/puma.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
```
Comment (bjkeller) [inline legacy/components/anemone/test/dummy/config/puma.rb:1]: I'd prefer not to have this here. Are you guys using it? And, if so, how?

**Review by bjkeller — CHANGES_REQUESTED** [review]
There are a couple of things I might be inclined to refactor, but I want to do some investigation first. I think the only changes I specifically mentioned are eslint warnings, a spelling error.
I'd also like to take out the legacy directory if we can.

File: frontend/src/components/sampleTypes/SampeTypeDefinitionForm.jsx [hunk 6]
```diff
@@ -0,0 +1,374 @@
+import { makeStyles } from '@material-ui/core';
```
[the remark this answers is not in the document]
Comment (marikoja) [reply frontend/src/components/sampleTypes/SampeTypeDefinitionForm.jsx:1]: I agree its a mouth full, updated to **SampleTypeForm**

File: legacy/components/anemone/test/dummy/config/puma.rb (hunk 5 above)
Comment (marikoja) [reply legacy/components/anemone/test/dummy/config/puma.rb:1]: I'm not using it, or actively using/modifying any of the ruby files.

**Review by bjkeller — APPROVED** [review]

**Pull request merged**